Feature/1 make a validation for recurring classes

Saeka03 opened the pull request:
## Overview
- Add validation for recurring classes.
- Use local time on the frontend UTC time on the backend.

## Review point
1. Run dummy.ts
2. Customer A changes one of the recurring classes. (e.g., Bob changes it to Fri 1:00 am PST)
3. Customer A cancels one or two classes and rebooks a new class. (e.g., Bob rebooks a class for Sun 23:00 PST)
4. Customer B changes one of the recurring classes on the same day and time as No.3. (Alice register a regular class Sun 23:00 PST)
 -> Check if the classes that Customer A booked on No.3 can't be booked and are canceled by instructors. 
5. Customer B changes one of the recurring classes to the date and time when the instructor is unavailable. (Helene Sun 23:00 or 23:30 PST)
 -> Check if the instructor's unavailability classes can't be booked and are canceled by instructors. 

- Ensure that local time is used on the regular classes page.
- Ensure that UTC time is stored in the DB.

## Consideration
- It seems that summertime starts from 2025 Mar 9 Sun, so after this date, time will be 1 hour ahead. (Only PST?)
- If a customer has some classes that can't be booked when they change a regular class, there is no problem when they change the regular class again if they don't book a new class. But if they book a new class, they will get an extra class, so we need to ask Haru-san about it.

Shingo023 reviewed (approved):
Thank you so much for working on one of the most challenging parts in our project! I'm sharing my feedback.

1. I reviewed your Review Points, and everything worked exactly as you explained!
2. I left some comments in your code. Some of them might not make sense—if so, feel free to ignore them.
3. Your second consideration seems quite serious. We definitely need to address it. 
4. If a class starts late (e.g., Wed 23:30), today's class (e.g., Wed Feb 26) is included in regular classes even if tomorrow's date (e.g., Thu Feb 27) is selected as the 'StartDate.' This is probably because I'm in Vancouver. However, since most customers are in Japan, this shouldn't be a major issue.
5. Bob only has one child, Emily, but Peppa is also being displayed.
![Screen Shot 2025-02-25 at 6 25 02 AM](https://github.com/user-attachments/assets/86242c3c-eb08-4e68-8d5a-ef03888774b5)

mtkksk1780 reviewed (approved):
Thank you for your hard work! Since @Shingo023 has reviewed this PR in detail, I trust his comments(Thank you Shingo!). I have left some comments regarding one of "Consideration."

> - It seems that summertime starts from 2025 Mar 9 Sun, so after this date, time will be 1 hour ahead. (Only PST?)

I found that many countries observe daylight saving time. [Wikipedia](https://ja.wikipedia.org/wiki/%E5%90%84%E5%9B%BD%E3%81%AB%E3%81%8A%E3%81%91%E3%82%8B%E5%A4%8F%E6%99%82%E9%96%93)

I think handling daylight saving time for all timezones in the program is quite challenging. Currently, I have two options in mind:

1. Frontend displays only Japan time (UTC+9)
→ Some online English conversation platforms also display only Japan time (see Images 1 and 2).
2. Implement a timezone dropdown function.
→ Create a new table and register various timezones, including both standard and daylight saving time. The frontend will have a timezone selection feature as shown below.

        Japan Time (JST) – UTC+9
        Pacific Time (PT) – UTC-8 / UTC-7 (DST)
        Eastern Time (ET) – UTC-5 / UTC-4 (DST)
        Central European Time (CET) – UTC+1 / UTC+2 (DST)
        Greenwich Mean Time (GMT) – UTC+0 / UTC+1 (DST)
        China Time (CST) – UTC+8
        India Time (IST) – UTC+5:30
        Australian Eastern Time (AET) – UTC+10 / UTC+11 (DST)

[Images 1: DMM Eikaiwa]
<img width="844" alt="Screenshot 2025-03-03 at 2 15 11" src="https://github.com/user-attachments/assets/c189316c-db1a-4d4a-aa72-b73a7127df31" />

[Images 2: Weblio Eikaiwa]
<img width="1159" alt="Screenshot 2025-03-03 at 2 35 47" src="https://github.com/user-attachments/assets/c3f24163-5d7c-4b94-8dc6-6dde2691aefa" />

Saeka03 commented on backend/src/controllers/recurringClassesController.ts:
```diff
@@ -143,16 +151,43 @@ export const updateRecurringClassesController = async (
+  // If classStartDate is equal to or shorter than today, it shouldn't be executed.
+  if (new Date(LocalClassStartDateTime) <= new Date(utcToday)) {
+    return res.status(400).json({ message: "Invalid Start From Date" });
+  }
```
Somehow I was obsessed with using UTC, but you're right! I modified it using Japan time. Also, I adjusted the minimum date uses can select the frontend side.

Saeka03 commented on backend/src/controllers/recurringClassesController.ts:
```diff
@@ -143,16 +151,43 @@ export const updateRecurringClassesController = async (
+        ) {
+          throw new Error("Recurring class is already taken");
+        }
+      });
```
Thanks for your suggestion. It makes sense! Also, I modified it to display the error message.

Saeka03 commented on backend/src/controllers/recurringClassesController.ts:
```diff
@@ -226,12 +258,50 @@ export const updateRecurringClassesController = async (
+        const until = getFirstDateInMonths(firstClassDate, 3);
+        console.log(until);
         dateTimes = createDatesBetween(firstClassDate, until);
+        console.log(dateTimes);
```
Thanks for catching it. You're right!

Saeka03 commented on frontend/src/app/customers/[id]/regular-classes/page.tsx:
```diff
@@ -15,9 +15,14 @@ function Page({ params }: { params: { id: string } }) {
+      toast.success(successMessage);
+    }
+
+    const warningMessage = searchParams.get("warningMessage");
+    if (warningMessage) {
+      toast.warning(warningMessage);
```
That's right, let's discuss it near future

Saeka03 commented on frontend/src/app/components/customers-dashboard/regular-classes/EditRegularClassForm.tsx:
```diff
@@ -142,11 +143,17 @@ function EditRegularClassForm({
-        : `/customers/${customerId}/regular-classes?message=${encodeURIComponent(data.message)}`;
-
-      router.push(targetURL);
+      if (data.messages[1]) {
+        const targetURL = isAdminAuthenticated
+          ? `/admins/customer-list/${customerId}?successMessage=${encodeURIComponent(data.messages[0])}&warningMessage=${encodeURIComponent(data.messages[1])}`
+          : `/customers/${customerId}/regular-classes?successMessage=${encodeURIComponent(data.messages[0])}&warningMessage=${encodeURIComponent(data.messages[1])}`;
+        router.push(targetURL);
```
Let me handle this in a different branch

Saeka03 commented:
@Shingo023 
Thanks for reviewing! (And sorry for the late check)

> 4. If a class starts late (e.g., Wed 23:30), today's class (e.g., Wed Feb 26) is included in regular classes even if tomorrow's date (e.g., Thu Feb 27) is selected as the 'StartDate.' This is probably because I'm in Vancouver. However, since most customers are in Japan, this shouldn't be a major issue.

 -> Thank you for pointing it out! You're right. It would be a bit problematic if users live in other countries. So, I fixed that.

> Bob only has one child, Emily, but Peppa is also being displayed.

 -> Sorry, I wrote the wrong data in the dummy.ts. It was my fault and I modified it.


@mtkksk1780 
Thank you for giving me the solutions! Yeah, it's quite challenging to deal with daylight saving time. And I decided to make another issue and handle the current bug related to it. I have one more option to use a library to handle daylight saving time, but what do you think?

mtkksk1780 commented:
@Saeka03 

> @mtkksk1780
> Thank you for giving me the solutions! Yeah, it's quite challenging to deal with daylight saving time. And I decided to make another issue and handle the current bug related to it. I have one more option to use a library to handle daylight saving time, but what do you think?
> 

(I'm sorry for the late reply)
I'm not sure which library you will use for the logic, but feel free to go with it if it works efficiently! From my current understanding, the logic can be implemented purely through calculations (but needs some experiments). However, I still don't have a clear overall picture, so go ahead with your approach first. If you get stuck, let's discuss it with everyone!
